internal/lint: test assumes a program exists in $PATH

Was running the tests and ran into this one:

```
--- FAIL: TestLint (0.72s)
    --- FAIL: TestLint/TestGolint (0.00s)
        lint_test.go:30: 
            	Error Trace:	lint_test.go:30
            	           				lint_test.go:67
            	Error:      	Received unexpected error:
            	           	exec: "golint": executable file not found in $PATH
            	Test:       	TestLint/TestGolint
FAIL
FAIL	github.com/cockroachdb/pebble/internal/lint	5.769s
```

The proper way to write this test would be to have it require a specific version of the linter, and do its own thing to run it, such as `go run golang.org/x/lint@some-commit args...`.

Alternatively, the bare minimum would be to skip the test if `golint` is not installed. That would still mean people could have different versions, though.

Also, beware that golint has been deprecated for some time: https://github.com/golang/go/issues/38968

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -30,15 +30,6 @@ jobs:
       with:
         go-version: ${{ matrix.go }}
 
-    - name: install crlfmt
-      run: go install github.com/cockroachdb/crlfmt@b3eff0b
-
-    - name: install golint
-      run: go install golang.org/x/lint/golint@6edffad
-
-    - name: install staticcheck
-      run: go install honnef.co/go/tools/cmd/staticcheck@2021.1.1
-
     - run: make test generate
 
   linux-crossversion:
@@ -65,15 +56,6 @@ jobs:
       with:
         go-version: "1.17"
 
-    - name: install crlfmt
-      run: go install github.com/cockroachdb/crlfmt@b3eff0b
-
-    - name: install golint
-      run: go install golang.org/x/lint/golint@6edffad
-
-    - name: install staticcheck
-      run: go install honnef.co/go/tools/cmd/staticcheck@2021.1.1
-
     - run: make testrace TAGS=
 
   linux-no-invariants:
@@ -86,15 +68,6 @@ jobs:
       uses: actions/setup-go@v2
       with:
         go-version: "1.17"
-
-    - name: install crlfmt
-      run: go install github.com/cockroachdb/crlfmt@b3eff0b
-
-    - name: install golint
-      run: go install golang.org/x/lint/golint@6edffad
-
-    - name: install staticcheck
-      run: go install honnef.co/go/tools/cmd/staticcheck@2021.1.1
 
     - run: make test TAGS=
 
@@ -109,15 +82,6 @@ jobs:
       with:
         go-version: "1.17"
 
-    - name: install crlfmt
-      run: go install github.com/cockroachdb/crlfmt@b3eff0b
-
-    - name: install golint
-      run: go install golang.org/x/lint/golint@6edffad
-
-    - name: install staticcheck
-      run: go install honnef.co/go/tools/cmd/staticcheck@2021.1.1
-
     - run: CGO_ENABLED=0 make test TAGS=
 
   darwin:
@@ -130,15 +94,6 @@ jobs:
       uses: actions/setup-go@v2
       with:
         go-version: "1.17"
-
-    - name: install crlfmt
-      run: go install github.com/cockroachdb/crlfmt@b3eff0b
-
-    - name: install golint
-      run: go install golang.org/x/lint/golint@6edffad
-
-    - name: install staticcheck
-      run: go install honnef.co/go/tools/cmd/staticcheck@2021.1.1
 
     - run: make test
 

--- a/go.mod
+++ b/go.mod
@@ -8,6 +8,7 @@ require (
 	github.com/cockroachdb/redact v1.0.8
 	github.com/ghemawat/stream v0.0.0-20171120220530-696b145b53b9
 	github.com/golang/snappy v0.0.3
+	github.com/hashicorp/go-version v1.6.0
 	github.com/klauspost/compress v1.11.7
 	github.com/kr/pretty v0.1.0
 	github.com/pmezard/go-difflib v1.0.0

--- a/internal/lint/lint_test.go
+++ b/internal/lint/lint_test.go
@@ -16,7 +16,15 @@ import (
 
 	"github.com/cockroachdb/errors"
 	"github.com/ghemawat/stream"
+	"github.com/hashicorp/go-version"
 	"github.com/stretchr/testify/require"
+)
+
+const (
+	cmdGo       = "go"
+	golint      = "golang.org/x/lint/golint@6edffad5e6160f5949cdefc81710b2706fbcd4f6"
+	staticcheck = "honnef.co/go/tools/cmd/staticcheck@2022.1.2"
+	crlfmt      = "github.com/cockroachdb/crlfmt@b3eff0b"
 )
 
 func dirCmd(t *testing.T, dir string, name string, args ...string) stream.Filter {
@@ -59,13 +67,23 @@ func TestLint(t *testing.T) {
 	}
 
 	t.Run("TestGolint", func(t *testing.T) {
+		// Go versions less than 1.17 do not support the `go run path@version`
+		// syntax.
+		// TODO(travers): This can be removed when support for go1.16 is dropped.
+		if goVersionMatches("< 1.17") {
+			t.Skip("go run path@version unsupported in versions < go1.17")
+		}
 		t.Parallel()
+
+		args := []string{"run", golint}
+		args = append(args, pkgs...)
 
 		// This is overkill right now, but provides a structure for filtering out
 		// lint errors we don't care about.
 		if err := stream.ForEach(
 			stream.Sequence(
-				dirCmd(t, pkg.Dir, "golint", pkgs...),
+				dirCmd(t, pkg.Dir, cmdGo, args...),
+				stream.GrepNot("go: downloading"),
 			), func(s string) {
 				t.Errorf("\n%s", s)
 			}); err != nil {
@@ -74,14 +92,21 @@ func TestLint(t *testing.T) {
 	})
 
 	t.Run("TestStaticcheck", func(t *testing.T) {
-		if strings.HasPrefix(runtime.Version(), "go1.18") {
-			t.Skip("Go 1.18 not yet supported. See: https://github.com/dominikh/go-tools/issues/1166")
+		// Go versions less than 1.17 do not support the `go run path@version`
+		// syntax.
+		// TODO(travers): This can be removed when support for go1.16 is dropped.
+		if goVersionMatches("< 1.17") {
+			t.Skip("go run path@version unsupported in versions < go1.17")
 		}
 		t.Parallel()
 
+		args := []string{"run", staticcheck}
+		args = append(args, pkgs...)
+
 		if err := stream.ForEach(
 			stream.Sequence(
-				dirCmd(t, pkg.Dir, "staticcheck", pkgs...),
+				dirCmd(t, pkg.Dir, cmdGo, args...),
+				stream.GrepNot("go: downloading"),
 			), func(s string) {
 				t.Errorf("\n%s", s)
 			}); err != nil {
@@ -213,13 +238,21 @@ func TestLint(t *testing.T) {
 	})
 
 	t.Run("TestCrlfmt", func(t *testing.T) {
+		// Go versions less than 1.17 do not support the `go run path@version`
+		// syntax.
+		// TODO(travers): This can be removed when support for go1.16 is dropped.
+		if goVersionMatches("< 1.17") {
+			t.Skip("go run path@version unsupported in versions < go1.17")
+		}
 		t.Parallel()
 
-		name := "crlfmt"
-		args := []string{"-fast", "-tab", "2", "-ignore", "^vendor/", "."}
+		args := []string{"run", crlfmt, "-fast", "-tab", "2", "-ignore", "^vendor/", "."}
 		var buf bytes.Buffer
 		if err := stream.ForEach(
-			stream.Sequence(dirCmd(t, pkg.Dir, name, args...)),
+			stream.Sequence(
+				dirCmd(t, pkg.Dir, cmdGo, args...),
+				stream.GrepNot("go: downloading"),
+			),
 			func(s string) {
 				fmt.Fprintln(&buf, s)
 			}); err != nil {
@@ -231,7 +264,7 @@ func TestLint(t *testing.T) {
 		}
 
 		if t.Failed() {
-			reWriteCmd := []string{name, "-w"}
+			reWriteCmd := []string{crlfmt, "-w"}
 			reWriteCmd = append(reWriteCmd, args...)
 			t.Logf("run the following to fix your formatting:\n"+
 				"\n%s\n\n"+
@@ -269,4 +302,17 @@ func lintIgnore(ignore string) stream.FilterFunc {
 		}
 		return nil
 	}
+}
+
+// goVersionMatches returns true if the Go runtime versions matches the given
+// semver constraint. If the runtime version does not contain a semver string
+// (i.e. it is a SHA), this function returns false.
+func goVersionMatches(constraint string) bool {
+	runtimeVersion := strings.TrimPrefix(runtime.Version(), "go")
+	goV, err := version.NewSemver(runtimeVersion)
+	if err != nil {
+		return false // Fail open.
+	}
+	c := version.MustConstraints(version.NewConstraint(constraint))
+	return c.Check(goV)
 }

